design/proposals/ambient-namespaces.md

```diff
@@ -0,0 +1,52 @@
+# Ambient Namespaces
+
+Status: Pending
+
+Version: Alpha
+
+## Abstract
+
+We want to introduce functionality to enable the use of environment metadata (ex: cluster namespace, Server URI) in prototype and component manifest files.
+
+## Motivation
+
+* Environments have configured metadata, such as the cluster’s namespace or server URI. Users [have requested](https://github.com/ksonnet/ksonnet/issues/222) that these fields be referenceable in both prototypes and component files.
+* Component and prototype manifests should not reference a specific environment's details. See Goals and Non Goals for more clarity on what "specific" implies.
+
+Example:
+
+Consider the following [nginx prototype](https://github.com/ksonnet/parts/blob/9d78d6bb445d530d5b927656d2293d4f12654608/incubator/nginx/examples/nginx.jsonnet) where the namespace is currently hardcoded:
+
+```jsonnet
+local k = import 'k.libsonnet';
+local nginx = import '../nginx.libsonnet';
+
+local namespace = "dev-alex";
+local appName = "nginx-app";
+
+k.core.v1.list.new([
+  nginx.parts.deployment.withServerBlock(namespace, appName),
+  nginx.parts.service(namespace, appName),
+  nginx.parts.serverBlockConfigMap(namespace, appName),
+])
+```
+ 
+This example does not lend well to the ksonnet desired model where components are environment agnostic; meaning a component should not need information on which environment the resulting Kubernetes Objects will be deployed to. Yet, it is unavoidable that these component / prototype configurations may need to reference a namespace, server URI, or other metadata.
+
+## Goals
+
+* Encourage behavior for component and prototype manifests to be environment-agnostic. 
+  Example: It is encouraged to write `nginx.parts.service(env.namespace, appName)`, but not `local namespace = "foo"` or `if (env.namespace == "foo") ... else ...`.
+* Component and prototype manifests should be able to refer to environment metadata fields, such as the environment namespace in the form of `env.namespace`.
+* Environment fields should be evaluated at run-time depending on which environment the user is trying to deploy to. 
+  Example: If the user had a `prod` environment configured to the namespace `prod-ns`, running `ks apply prod` should evaluate `env.namespace` to the value `prod-ns`.
+
+## Non Goals
+
+The following are related but not direct goals of this specific proposal
+
+* Rewriting existing prototypes to remove hard-coded namespace values.
```

Comment:
Ha, I think this should be an explicit goal, FWIW. That they are hard-coded is the reason we _aren't_ using ambient namespaces to begin with, right? (Since ambient namespaces are given to objects that _don't_ have the `namespace` set.)

Reply:
We probably have to supply the rewrites as different versions, otherwise we'll run into backwards-compatibility problems when a 0.8.0 ks user attempts to use these updates.

Reply:
Yep, that makes sense.